skorch/net.py

```diff
@@ -725,43 +725,41 @@ def fit_loop(self, X, y=None, epochs=None, **fit_params):
             'dataset_valid': dataset_valid,
         }
 
-        y_train_is_ph = uses_placeholder_y(dataset_train)
-        y_valid_is_ph = uses_placeholder_y(dataset_valid)
-
-        for _ in range(epochs):
+        for epoch in range(epochs):
             self.notify('on_epoch_begin', **on_epoch_kwargs)
 
-            train_batch_count = 0
-            for data in self.get_iterator(dataset_train, training=True):
-                Xi, yi = unpack_data(data)
-                yi_res = yi if not y_train_is_ph else None
-                self.notify('on_batch_begin', X=Xi, y=yi_res, training=True)
-                step = self.train_step(Xi, yi, **fit_params)
-                train_batch_count += 1
-                self.history.record_batch('train_loss', step['loss'].item())
-                self.history.record_batch('train_batch_size', get_len(Xi))
-                self.notify('on_batch_end', X=Xi, y=yi_res, training=True, **step)
-            self.history.record("train_batch_count", train_batch_count)
-
-            if dataset_valid is None:
-                self.notify('on_epoch_end', **on_epoch_kwargs)
-                continue
+            self._single_epoch(dataset_train, training=True, epoch=epoch, **fit_params)
 
-            valid_batch_count = 0
-            for data in self.get_iterator(dataset_valid, training=False):
-                Xi, yi = unpack_data(data)
-                yi_res = yi if not y_valid_is_ph else None
-                self.notify('on_batch_begin', X=Xi, y=yi_res, training=False)
-                step = self.validation_step(Xi, yi, **fit_params)
-                valid_batch_count += 1
-                self.history.record_batch('valid_loss', step['loss'].item())
-                self.history.record_batch('valid_batch_size', get_len(Xi))
-                self.notify('on_batch_end', X=Xi, y=yi_res, training=False, **step)
-            self.history.record("valid_batch_count", valid_batch_count)
-
-            self.notify('on_epoch_end', **on_epoch_kwargs)
+            if dataset_valid is not None:
+                self._single_epoch(dataset_valid, training=False, epoch=epoch, **fit_params)
+
+            self.notify("on_epoch_end", **on_epoch_kwargs)
         return self
 
+    def _single_epoch(self, dataset, training, epoch, **fit_params):
```

Comment:
We can update this signature to

```py
def _single_epoch(self, dataset, training, prefix, 
                  step_fn, epoch, **fit_params):
    ...
```

And let the caller pass the arguments in:

```py
 for epoch in range(epochs):
    self.notify('on_epoch_begin', **on_epoch_kwargs)
    self._single_epoch(dataset_train, training=True, prefix="train",
					   step_fn=self.train_step, epoch=epoch, **fit_params)

    if dataset_valid is not None:
        self._single_epoch(dataset_valid, training=False, prefix="valid",
					       step_fn=self.validation_step, epoch=epoch, **fit_params)
    self.notify("on_epoch_end", **on_epoch_kwargs)
```